### PR TITLE
Add response callback to ChatBox

### PR DESCRIPTION
**Pull request by ahuang11**
Instead of having a separate Interface, like [here](https://github.com/holoviz/panel/pull/5313), this supports general callbacks and is more customizable with yields.

image

```python
import time
import panel as pn
pn.extension()

async def echo_callback(message: str, interface: pn.widgets.ChatBox):
    yield pn.indicators.LoadingSpinner(value=True, width=17, height=17)
    
    time.sleep(0.75)
    message_echo = ""
    for character in message:
        time.sleep(0.05)
        message_echo += character
        yield message_echo

interface = pn.widgets.ChatBox(response_callback=echo_callback)
interface.servable()
```

```python
import asyncio
import random
import panel as pn
import openai
pn.extension()

async def openai_callback(message: str, chat_box: pn.widgets.ChatBox):
    yield pn.indicators.LoadingSpinner(value=True, width=17, height=17)

    response = openai.ChatCompletion.create(
        model="gpt-3.5-turbo",
        messages=[{"role": "user", "content": message}],
        temperature=0.9,
        max_tokens=100,
        stream=True
    )
    message = ""
    for chunk in response:
        message += chunk["choices"][0]["delta"].get("content", "")
        yield message

interface = pn.widgets.ChatBox(response_callback=openai_callback)
interface.servable()
```

https://user-images.githubusercontent.com/15331990/255759560-12c46f66-aaf1-453b-91f2-2d687995a0ee.png

### CODE REVIEW

**Comment by ahuang11** [comment]
If users prefer not to use async, they can use the `response_placeholder` like:
```python
import time
import panel as pn
pn.extension()

def echo_callback(message: str, interface: pn.widgets.ChatBox):
    time.sleep(0.75)
    return message

interface = pn.widgets.ChatBox(response_callback=echo_callback, response_placeholder="Loading...")
interface.servable()
```


https://github.com/holoviz/panel/assets/15331990/32c60ff0-8de9-4828-95a6-a494bf63292d

**Comment by ahuang11** [comment]
The video demos:
1. user uploading a CSV file
2. internally extracts the column names
3. asks OpenAI to write some code with hvplot to plot
4. the plot gets shown and is interactive

https://github.com/holoviz/panel/assets/15331990/65b5280f-5613-4f54-bcba-0e286dddb5f6

File: panel/widgets/chatbox.py [hunk 1]
```diff
@@ -219,9 +302,23 @@ class ChatBox(CompositeWidget):
         e.g. `[{'You': ['Welcome!', 'Good bye!']}]` The message(s) can be
         any Python object that can be rendered by Panel.""")
 
-    primary_name = param.String(default=None, doc="""Name of the primary user (the one who inputs messages);
+    primary_name = param.String(default=None, doc="""
+        Name of the primary user (the one who inputs messages);
         the first key found in value will be used if unspecified.""")
 
+    response_callback = param.Callable(default=None, doc="""
```
Comment (MarcSkovMadsen) [inline panel/widgets/chatbox.py:309]: Please consider that right now everyone is in a flow of an "assistant" responding to a "user". Multiple users that should be responded to. Or specialized bots responding to another specialized bots.

I.e. I believe its important to be able to be able to support multiple callbacks. One per role/ name.

That is what I implemented in https://github.com/holoviz/panel/pull/5279.

**Comment by MarcSkovMadsen** [comment]
Please consider whether the `ChatBox` or the `ChatBox.value` should be the input to the callback. The latter will give more possibilities, but maybe the `value` is the right thing to provide.

`value` is what Gradio provides in their callback. Its also what I implemented here https://github.com/holoviz/panel/pull/5279. If I wanted to use the `ChatBox` in the callback I would define the function after the `ChatBox` was define and include it directly in the function.

**Comment by MarcSkovMadsen** [comment]
I really like this Path and I think hvplot is a good example for a chatbot because the interface is relative "simple". We can the the LLM be trained on a hvPlot docstring or provide that as context. We can provide simple examples to the LLM.

**Comment by MarcSkovMadsen** [comment]
Its also worth noting that now its possible to use transformers in pyodide via https://github.com/whitphx/transformers.js.py. That can give us nice examples to easily deploy.

**Comment by ahuang11** [comment]
> ```python
> 
> async def openai_callback(contents: Any, chat_box: pn.widgets.ChatBox):
> ```

It provides the ChatBox serialized row object, which admittedly is a bit hard to follow:
1. if the input ChatBox value is `{"You": [b"col1,col2,col3\n1,2,3"]}` with mime_type `text/csv`
2. it becomes a ChatRow object with a `[pn.pane.DataFrame]` list object (I regret implementing list as ChatRow's value type as it adds to the complexity; maybe we should deprecate `ChatBox` in favor of `ChatInterface`).
3. The `value` (if widget), `object` (if pane), `objects` (if layout) of the ChatRow object is then extracted and returned as the `contents` in the `response_callback` signature and if there's only one item, the list drops.
4. If users need more control / parameters, they can use the `chat_box` arg to access whatever they need

File: panel/widgets/chatbox.py (hunk 1 above)
Comment (ahuang11) [reply panel/widgets/chatbox.py:309]:
```
import re
from typing import Any

from panel.io.mime_render import exec_with_return
import pandas as pd
import panel as pn
import openai

pn.extension()

DATAFRAME_PROMPT = """
    Here are the columns in your DataFrame: {columns}.
    Create  a plot with hvplot that highlights an interesting
    relationship between the columns with hvplot groupby kwarg.
"""

CODE_REGEX = re.compile(r"```python(.*?)```", re.DOTALL)


async def respond_with_openai(contents: Any):
    # extract the DataFrame
    if isinstance(contents, pd.DataFrame):
        global df
        df = contents
        columns = contents.columns
        message = DATAFRAME_PROMPT.format(columns=columns)

    # ask OpenAI to plot
    response = await openai.ChatCompletion.acreate(
        model="gpt-3.5-turbo",
        messages=[{"role": "user", "content": message}],
        temperature=0,
        max_tokens=500,
        stream=True,
    )
    message = ""
    async for chunk in response:
        message += chunk["choices"][0]["delta"].get("content", "")
        yield {"OpenAI": message}


async def respond_with_executor(code: str):
    return {"Executor": exec_with_return(code=code, global_context=globals())}


async def response_callback(
    contents: Any,
    name: str,
    chat_box: pn.widgets.ChatBox,
):
    if name == "You":
        async for chunk in respond_with_openai(contents):
            yield chunk
    elif CODE_REGEX.search(contents):
        yield await respond_with_executor(CODE_REGEX.search(contents).group(1))


pn.widgets.ChatBox(
    message_input_widgets=[pn.widgets.FileInput()], response_callback=response_callback
).servable()
```

This semi-works, but the executor responds too fast before the OpenAI response is completely finished.

image

**Comment by ahuang11** [comment]
> Have you considered if its possible to seperately keep track of the (string) value from the response and the rendered value? I think that will be much easier to handle then having to guess the response value from the rendered value.

Yes I am planning to rewrite ChatRow and add ChatMessage:

```python
import param
import panel as pn
import datetime
pn.extension()

class ChatMessage(param.Parameterized):

    name = param.String(doc="Name of the user who sent the message.")

    value = param.Parameter(doc="The raw message content.")

    icon = param.String(default=None, doc="The icon to use for the user.")

    liked = param.Boolean(default=False, doc="Whether the message was liked.")

    timestamp = param.Date(
        default=datetime.datetime.utcnow(), doc="Timestamp of the message."
    )

    rendered = param.Parameter(doc="The rendered message content.")

class ChatEntry(param.Parameterized):

    chat_message = param.ClassSelector(
        class_=ChatMessage, doc="The ChatMessage to display."
    )
```

**Comment by ahuang11** [comment]
Philipp and I just discussed this and I think I'll deprecate ChatBox + ChatRow completely in favor of:

1. ChatInterface
2. ChatLog
3. ChatEntry
4. ChatMessage

Will share the API of each shortly~

File: panel/widgets/chatbox.py (hunk 1 above)
Comment (ahuang11) [reply panel/widgets/chatbox.py:313]: Okay, I got the multi responses working:

https://github.com/holoviz/panel/assets/15331990/e5269089-093c-42b0-b74d-cf23ed2bfc30

**Comment by ahuang11** [comment]
This is what I have so far; still thinking about the methods for each...

```python
import datetime
import param
from typing import BinaryIO
from panel.widgets import CompositeWidget, Widget
from panel.layout import Column, Row
from panel.reactive import ReactiveHTML


class ChatMessage(param.Parameterized):

    value = param.Parameter(doc="""
        The message contents. Can be a string, pane, widget, layout, etc.""")

    name = param.String(doc="Name of the user who sent the message.")

    icon = param.ClassSelector(class_=(str, BinaryIO), doc="""
        The icon to use for the user. Can be a URL, a file path, or a binary IO.""")

    liked = param.Boolean(default=False, doc="""
        Whether the message is favorited.""")

    timestamp = param.Date(default=datetime.datetime.utcnow(), doc="""
        Timestamp of the message. Defaults to the instantiation time.""")

    @property
    def json(self) -> dict:
        """
        Return the ChatMessage as JSON. For panes,
        extract the pane's underlying `object`.
        For widgets, extract the widget's `value`.
        For layouts, extract the layout's `objects`
        and recursively extract the children's
        `object` or `value`.
        e.g.
        ```
        {
            "value": "Hello, world!",
            "name": "User",
            "icon": "https://...",
            "liked": False,
            "timestamp": "2021-01-01T00:00:00.000Z",
        }
        ```
        """


class ChatEntry(ReactiveHTML):  # Pane?

    object = param.ClassSelector(
        class_=ChatMessage, doc="The ChatMessage to display."
    )

    show_name = param.Boolean(default=False, doc="Whether to show the name.")

    show_icon = param.Boolean(default=False, doc="Whether to show the icon.")
    
    show_liked = param.Boolean(default=False, doc="Whether to show the liked button.")

    show_timestamp = param.Boolean(default=False, doc="Whether to show the timestamp.")

    # reactive html will handle name, icon, and like updates
    _template = """
    """

    def __init__(self, object, **params):
        super().__init__(object=object, **params)

class ChatLog(Column):  # Layout?

    value = param.List(doc="The list of ChatEntry values.", item_type=ChatEntry)

    @property
    def messages(self):
        """
        Return the list of ChatMessage values.
        """
        ...

    def export(self, format: str = "json"):
        """
        Exports the chat log into a format that can be
        used in other applications.
        """
        ...
    
    def delete(self, index: int):
        """
        Deletes a message at the given index.
        """
        ...

    def reset(self):
        """
        Resets the chat log.
        """
        ...
    
    def replace(self, index: int, value: ChatMessage):
        """
        Replaces a message at the given index with the given value.
        """
        ...
    
    def insert(self, index: int, value: ChatMessage):
        """
        Inserts a message at the given index with the given value.
        """
        ...
    
    def append(self, value: ChatMessage):
        """
        Appends a message with the given value.
        """
        ...
    
    def prepend(self, value: ChatMessage):
        """
        Prepends a message with the given value.
        """
        ...

# Ideally, this would be a generalized widget
# that combines TextInput, TextAreaInput, FileInput, etc.
# Imagine GitHub's comment box.
class ChatInput(ReactiveHTML):  # Widget
    """
    Low level widget that allows users to either
    type a message, paste, or upload an attachment.
    """

    value = param.ClassSelector(class_=(str, bytes), doc="""
        The message contents. Can be a string or bytes.""")

    def reset():
        ...


class ChatInterface(CompositeWidget):
    """
    High level widget that contains the chat log and the chat input.
    """

    value = param.ClassSelector(class_=ChatLog, doc="""
        The ChatLog widget.""")
    
    input_widget = param.ClassSelector(class_=Widget, doc="""
        Widget to use for the input.""")

    primary_user = param.String(default="User", doc="""
        The name of the primary user.""")

    response_callback = param.Callable(doc="""
        Callback to execute when the user sends a message.""")

    response_placeholder = param.Parameter(doc="""
        Placeholder to display while the response callback is running.""")

    response_user = param.String(default="Assistant", doc="""
        The default name of the user who responds to the primary user.
        If the response callback returns a ChatMessage with a name,
        that name will be used instead.
        """)

    response_icon = param.ClassSelector(class_=(str, BinaryIO), doc="""
        The default icon to use for the user who responds to the primary user.
        Can be a URL, a file path, or a binary IO.""")

    def __init__(self, **params):
        super().__init__(**params)

class OpenAIChatInterface(ChatInterface):

    api_key = param.String(doc="""
        The OpenAI API key.""")
    
    model = param.String(doc="""
        The OpenAI model to use.""")
    
    temperature = param.Number(default=0.5, bounds=(0, 1), doc="""
        The temperature to use for the OpenAI model.""")
    
    max_tokens = param.Integer(default=100, bounds=(1, None), doc="""
        The maximum number of tokens to generate.""")
```

**Comment by ahuang11** [comment]
On second thought, it might make sense to call it ChatBlock instead of ChatLog, and add the response params in ChatBlock.

```python

class ChatBlock(Column):  # Layout?

    value = param.List(doc="The list of ChatEntry values.", item_type=ChatEntry)

    response_callback = param.Callable(doc="""
        Callback to execute when the user sends a message.""")

    response_placeholder = param.Parameter(doc="""
        Placeholder to display while the response callback is running.""")

    response_user = param.String(default="Assistant", doc="""
        The default name of the user who responds to the primary user.
        If the response callback returns a ChatMessage with a name,
        that name will be used instead.
        """)

    response_icon = param.ClassSelector(class_=(str, BinaryIO), doc="""
        The default icon to use for the user who responds to the primary user.
        Can be a URL, a file path, or a binary IO.""")

    @property
    def messages(self):
        """
        Return the list of ChatMessage values.
        """
        ...

    def export(self, format: str = "json"):
        """
        Exports the chat log into a format that can be
        used in other applications.
        """
        ...
    
    def delete(self, index: int):
        """
        Deletes a message at the given index.
        """
        ...

    def reset(self):
        """
        Resets the chat log.
        """
        ...
    
    def replace(self, index: int, value: ChatMessage):
        """
        Replaces a message at the given index with the given value.
        """
        ...
    
    def insert(self, index: int, value: ChatMessage):
        """
        Inserts a message at the given index with the given value.
        """
        ...
    
    def append(self, value: ChatMessage):
        """
        Appends a message with the given value.
        """
        ...
    
    def prepend(self, value: ChatMessage):
        """
        Prepends a message with the given value.
        """
        ...

# Ideally, this would be a generalized widget
# that combines TextInput, TextAreaInput, FileInput, etc.
# Imagine GitHub's comment box.
class ChatInput(ReactiveHTML):  # Widget
    """
    Low level widget that allows users to either
    type a message, paste, or upload an attachment.
    """

    value = param.ClassSelector(class_=(str, bytes), doc="""
        The message contents. Can be a string or bytes.""")

    def reset():
        ...


class ChatInterface(CompositeWidget):
    """
    High level widget that contains the chat log and the chat input.
    """

    value = param.ClassSelector(class_=ChatLog, doc="""
        The ChatLog widget.""")
    
    input_widget = param.ClassSelector(class_=Widget, doc="""
        Widget to use for the input.""")

    primary_user = param.String(default="User", doc="""
        The name of the primary user.""")

    def __init__(self, **params):
        super().__init__(**params)

```

**Comment by MarcSkovMadsen** [comment]
Would it be an idea to put the design proposal  (picture + class definitions) into the PR?

That would make it easier to discuss and iterate on. Plus the picture could be used later in the Reference Guides to give users an overview over the building blocks they can put together to create custom chat interfaces.

**Comment by ahuang11** [comment]
Sounds good.

**Comment by ahuang11** [comment]
I started a new PR [here](https://github.com/holoviz/panel/pull/5333)

Closing this in favor for that.
